### ISSUE TITLE
Keyboard handling in the Cheat Screen, SDL 3 

### ISSUE THREAD

**Issue by Dragon-Baroque** [exult/exult#531]
@DominusExult, @wench, I would like your advice.

The Cheat Screen keyboard handler takes the SDL KEY_DOWN events and dispatches on the SDL_Keycode ( the pressed key ) with almost no attention to the SDL_Keymod ( the Shift, Alt, ... modifiers ). It only looks at the modifiers to catch Alt-Ctl-S and take a snapshot like the Core of Exult does - and this is not Cheat Screen related -, and to respect the Upper case / Lower case characters when giving a name to an NPC.

The Cheat Screen handles the alphabetic letters A through Z, the digits 0 to 9, the special Return, Escape and Backspace, plus the three tricky keys Plus, Minus and Asterisk.
Those three are tricky because some are Shifted variants of other keys, so that ignoring the Shift modifier forces this code to hardcode the base key of each of the three keys.
But the current code assumes an **US keyboard** where Minus is base, but Asterisk is Shift 8, and Plus is Shift Equal.

It could well be how the Cheat Screen worked in the original Ultima VII games, but we should be able to do better, don't we ?

@DominusExult, when you use the Cheat Screen and need to enter one of these three keys, do you know what key you should press or do you use the Numeric Keypad instead ?

@Dominus, @wench, do you like enough never having to press Shift in the Cheat Screen - except for NPC names -, so that you would censor me if I proposed a change that requires you to press Shift appropriately ? This affects me too, since the French AZERTY layout has the digit keys Shifted, and I like - a bit - not having to press Shift in the Cheat Screen to reach the digits...

Note that I cannot satisfy the NLS layout compliance in full without the use of the  Shift key, because the German QWERTZ layout places Plus and Asterisk on the same key, with Shift to differentiate them, and the NPC menu needs both keys.

I am facing these choices and I would like your review :

1. Handle the keyboard Shift modifier always. This has the advantage of supporting all NLS layouts, but this doubles all the alphabetic cases in the code ( for lower case and upper case ). I have ideas on how to avoid that doubling, though. But this requires you to press Shift appropriately. This is my question above.
2. Never accept the three keys ( and possibly the digits too ) from the main part of the keyboard, only from the Numeric Keypad. I could handle the digit keys of the numeric keypad to return digits always, whether or not NumLock is set.
3. Leave the Cheat Screen as is, knowing that the three keys can be obtained from the numeric keypad for those of us who use a non US keyboard layout.

**Comment by DominusExult** [comment]
It's been a long time that I realized this problem :)
I'm always using a full keyboard with a numpad out of habit with former financial jobs and I don't like the cramped arrow/page up/down keys...
Weirdly the German `+` key not on the numpad is not properly registering even though it doesn't need a modifier key: In the entry screen you can set the time rate with `+` and `-` , instead of upping the rate it will enter `+` in the `Select `prompt and claim it is an invalid key :) (the `-` key works correctly, though).

Looking at the US layout, these keys also need the shift modifier to work AFAICT. 
So my guess is that option 1 is the way to go, especially as full keyboards have gone out of fashion (sort of).

**Comment by wench** [comment]
This is not an easy thing to fix. Unless SDL can tell us what the Translated modified character is I'm not seeing a way to do it properly. If we can get the proper translated shifted character I would choose option 1,

The only real problem codes being +-* gives some options, we could change those to instead use keys that are unmodified for all layouts. I never considered layouts other than US when I first wrote all this + and - could be changed to cursor keys. For * i have no idea just some other unused key in the specific menu. 

the German + doesn't work because the code explicitly checks for = or KP+ being pressed and doesn't check for +.

**Comment by Dragon-Baroque** [comment]
> Unless SDL can tell us what the Translated modified character is I'm not seeing a way to do it properly.

I cannot tell about SDL 2. But SDL 3 publishes a `SDL_GetKeyFromScancode( SDL_Scancode, SDL_Keymod )`.
If you are in a KeyDown event, then the `SDL_Scancode` and the `SDL_Keymod` are available, and you can play with `SDL_GetKeyFromScancode` with the SDL_Keymod of the event suitably modified to your needs.

The SDL_Keycode of a KeyDown event ( forgetting the exceptions of the French digits ) is a `SDL_GetKeyFromScancode` with the `SDL_Keymod` argument set to 0. But in the KeyDown event handler, you can ( and this is the recommended way ) call again `SDL_GetKeyFromScancode` with the `SDL_Keymod` you need.

In other words, I think that obtaining the Translated modified character is
```
SDL_GetKeyFromScancode( event.key.scancode, event.key.mod & ( SDL_KMOD_SHIFT | SDL_KMOD_CAPS ) )
```
if you want to avoid the interference of Alt / Alt Gr / Mode, or a full
```
SDL_GetKeyFromScancode( event.key.scancode, event.key.mod )
```
if you want the complete Keymap.

**Comment by Dragon-Baroque** [comment]
> + and - could be changed to cursor keys. For * i have no idea just some other unused key in the specific menu.

If Plus becomes Right, and Minus becomes Left, what about Home for Asterisk ?

**Comment by DominusExult** [comment]
Home/end is problematic on macOS (as it's the fn key + arrows up/down).
but maybe just cursor up for *?
Problem we have no arrow symbols in fonts.vga

**Comment by wench** [comment]
Cheatscreen uses Mainshp.flx:9 for it's font but still same prblem no arrows
However Left and right arrows are easy to simulate just draw overlapping characters < with - for left and - with > for right. For up it looks like i with ^ would work well enough. There doesn't seem to be a way to make a down arrow

I'll try this change in the next day or so

**Comment by wench** [comment]
Played around a bit trying to draw arrows by combining characters. This is about as good as i can do. I had to draw some additional black lines to tidy things up a bit

Annoyingly the font has a slightly thicker lines SI than BG but I think I have it looking good enough

BG:
![arrowsbg](https://github.com/exult/exult/assets/1016583/5f0f8ce5-6fa3-4300-8254-84ba8c0ebfbb)


SI:
![arrowssi](https://github.com/exult/exult/assets/1016583/378a718e-78ac-479d-be62-8928f0e4650f)

I'l get this stuff in tomorrow or so

**Comment by DominusExult** [comment]
That looks good!

**Comment by DominusExult** [comment]
@wench as we talked about the `[E]xit` vs `[X]it` in the palette screen, I would like to change throughout the cheat screen everything to `[ESC] Exit` and also change the input cancel message (`-1 to cancel`) to `ESC to cancel`.
I will do that when you have done the `+ - *`  replacement, so don't bother doing it. My only question is whether to keep the option to enter `-1` around, IMO it can go away.

**Comment by wench** [comment]
I would get rid of the -1 stuff

**Comment by wench** [comment]
Pushed the arow changes. I hope I replaced all cases

**Comment by DominusExult** [comment]
I just noticed that I will have to do something special for the mobile platforms (iOS and Android). The keyboards are without arrow keys :(
So I may have to ifdef these cases. This sucks a bit

**Comment by wench** [comment]
I had a though of adding hot spots for mouse/touch. It would probably be a better way of doing things than needing on screen keyboard input.

**Comment by Dragon-Baroque** [comment]
> Pushed the arrow changes. I hope I replaced all cases.

I think yes. And I have migrated it to SDL 3 in my branch `exult-sdl3` without any problem.

**Comment by wench** [comment]
Are there any considerations I should make implementing mouse and touch support

**Comment by DominusExult** [comment]
> > Pushed the arrow changes. I hope I replaced all cases.
> 
> I think yes. And I have migrated it to SDL 3 in my branch `exult-sdl3` without any problem.

I just tested your SDL3 branch and there is a problem: all input expects capital letters. So in the first screen `t` does not work, but `T` (`shift-t`) works. It should prefer the lower case. none-shift characters.

> Are there any considerations I should make implementing mouse and touch support

only thing I can think of right now is that a single touch toggles the onscreen keyboard on iOS and Android (SDL_StopTextInput & SDL_StartTextInput)

**Comment by Dragon-Baroque** [comment]
> I just tested your SDL3 branch and there is a problem: all input expects capital letters. So in the first screen `t` does not work, but `T` (`shift-t`) works. It should prefer the lower case. none-shift characters.

Is your SDL 3 build recent ? In the past few days there has been a big rework and `SDLK_T` now means `t` not `T`, whereas `SDLK_t` does no longer exist.

**Comment by DominusExult** [comment]
> > I just tested your SDL3 branch and there is a problem: all input expects capital letters. So in the first screen `t` does not work, but `T` (`shift-t`) works. It should prefer the lower case. none-shift characters.
> 
> Is your SDL 3 build recent ? In the past few days there has been a big rework and `SDLK_T` now means `t` not `T`, whereas `SDLK_t` does no longer exist.

oops :)

**Comment by Dragon-Baroque** [comment]
But this is a good point.
With the three `+`, `-`, and `*` out of the way thanks to Wench, I should revert to make the main part of the Cheat menu - NPC names excepted - case insensitive.
As of now `[T] Teleport` accepts `t` but not `T`, this is not ideal.

**Comment by DominusExult** [comment]
With semi current SDL3 (of course the last few SD3 commits broke something else) the lower keys work fine(and upper case does not as you pointed out).

**Comment by Dragon-Baroque** [comment]
Refreshed my branches, with 
- Case insensitive in the Cheat Screen except for the NPC name,
- Latest changes on SDL 3,
- Latest changes on Exult master.

**Comment by DominusExult** [comment]
@Dragon-Baroque everything works fine now (except for the following but that has nothing to do with SDL3)

@wench your change to cancel everything via ESC somehow breaks the Set Time cheat in the main screen. Once you cancel setting the time via ESC, the cheat screen accepts no more input (but AFAICT Exult is not hanging).
This is with f15170fff - the first change to cancel via ESC.

**Comment by wench** [comment]
ok fixed Set Time and tweaked how escaping works. Escaping from an input mode will now switch to the cancelled mode showing a canceled message. Escape can now be used to clear modes that ask to press a key

@DominusExult a question for you, how does iOS handle escaping at the moment? Does it handle it at all

**Comment by DominusExult** [comment]
Thank you!
On iOS (and Android) we most always have the overlay ESC button in the lower left corner. In the cgeat screen it is shown and useable when the onscreen keyboard is toggled off.
Screenshot (on Android a bit different looking):
![IMG_1982](https://github.com/exult/exult/assets/3313896/f82bfa42-7b33-4a0c-8b09-6a09a5123f89)

**Comment by wench** [comment]
Ok cool. My plan is to make all the [key] bits clickable/tappable. That screenshot was helpful to see that the hotspots would not be too small. They would be about the same size as that esc button

**Comment by Dragon-Baroque** [comment]
> ok fixed Set Time and tweaked how escaping works. Escaping from an input mode will now switch to the canceled mode showing a canceled message.

Thank you. My branch `exult-sdl3` is now rebased onto that fix. ( I leave the GTK 4 branches somewhat behind, lesser priority ). You can use either SDL 3 as of yesterday - easier, I guess - or as of today _but without the commit 4fa519682, which reverts the changes in SDL_pixels.h_.

> Ok cool. My plan is to make all the [key] bits clickable/tappable.

Please, what kind of SDL ( SDL 2 ) events will that produce ? Just to have get an idea on how they should be migrated to SDL 3.

**Comment by wench** [comment]
None. It will just be an addition to the shared_input method where clicks ion the hotspots will set the command variable with the simulated keypress.

**Comment by Dragon-Baroque** [comment]
> with the simulated _**keypress.**_

Easy then. Thank you.

**Comment by wench** [comment]
Changed my mind a little and what I am doing is like this
```
				SDL_KeyCode simulate_key = SDLK_UNKNOWN;
				...
				if (simulate_key) {
					event.type           = SDL_KEYDOWN;
					event.key.keysym.sym = simulate_key;
				}
```
The ... is all the mouse and touch input code mostly copied from `Gump_manager::handle_modal_gump_event`

**Comment by wench** [comment]
I setup an Android build environment to test things out on my phone. It became very obvious very quickly that Hotspots for the left right arrows are not ideal when trying to do things like scrolling through nps. I'm going to implement swiping for these as that seems like a more natural control method for phones.

**Comment by DominusExult** [comment]
I think this one can be closed. The swiping has been implemented as far as I can tell from the Android emulator

**Comment by DominusExult** [comment]
Will have to reopen. The cheat screen on iOS is not properly aligning touch input with where it touches on the cheat screen.

**Comment by wench** [comment]
If it is iOS only does iOS use hidpi mode?

**Comment by DominusExult** [comment]
It seems to be only affecting iOS and only on SDL3 (the 1.12 build works fine).
We no longer have the enable/disable HiDPi toggle in Exult as this handled directly by SDL3 now for iOS. I remember that the cheat screen needed special handling for hidpi at some point. I'll need to investigate and set my mind to it again. And it really only affects the cheat screen.

**Comment by DominusExult** [comment]
Solution is adding `SDL_ConvertEventToRenderCoordinates(renderer, &event);` to `case SDL_EVENT_MOUSE_BUTTON_DOWN: {` as is done elsewhere in our code.
